synapse/lib/storm.py

```diff
@@ -783,6 +784,80 @@
     },
 )
 
+class DmonManager(s_base.Base):
+    '''
+    Manager for StormDmon objects.
+    '''
+    async def __anit__(self, core):
+        await s_base.Base.__anit__(self)
+        self.core = core
+        self.dmons = {}
+        self.enabled = False
+
+        self.onfini(self._stopAllDmons)
+
+    async def _stopAllDmons(self):
+        await asyncio.gather(*[dmon.fini() for dmon in self.dmons.values()])
+
+    async def addDmon(self, iden, ddef):
+        dmon = await StormDmon.anit(self.core, iden, ddef)
+        self.dmons[iden] = dmon
+        # TODO Remove default=True when dmon enabled CRUD is implemented
+        if self.enabled and ddef.get('enabled', True):
+            await dmon.run()
+        return dmon
+
+    def getDmonRunlog(self, iden):
+        dmon = self.dmons.get(iden)
+        if dmon is not None:
+            return dmon._getRunLog()
+        return ()
+
+    def getDmon(self, iden):
+        return self.dmons.get(iden)
+
+    def getDmonPacked(self, iden):
+        dmon = self.dmons.get(iden)
+        if dmon:
+            return dmon.pack()
+
+    def getDmonsPacked(self):
+        return list(d.pack() for d in self.dmons.values())
+
+    async def popDmon(self, iden):
+        '''Remove the dmon and fini it if its exists.'''
+        logger.debug(f'Poping dmon {iden}')
+        dmon = self.dmons.pop(iden, None)
+        if dmon:
+            await dmon.fini()
+
+    async def start(self):
+        '''
+        Start all the dmons.
+        '''
+        if self.enabled:
+            return
+        for dmon in list(self.dmons.values()):
```

Comment:
We have a turn-on/turn-off/turn-on problem here where the underlying dmon objects will ahve been fini'd